src/anomalib/metrics/spro.py

```diff
@@ -0,0 +1,158 @@
+
+    def __init__(self, threshold: float = 0.5, saturation_config: dict | None = None, **kwargs) -> None:
+        super().__init__(**kwargs)
+        self.threshold = threshold
+        self.saturation_config = saturation_config
+        self.add_state("preds", default=[], dist_reduce_fx="cat")
+        self.add_state("targets", default=[], dist_reduce_fx="cat")
+
+    def update(self, predictions: torch.Tensor, _: torch.Tensor, masks: torch.Tensor) -> None:
+        """Compute the SPRO score for the current batch.
+
+        Args:
+            predictions (torch.Tensor): Predicted anomaly masks
+            _ (torch.Tensor): Unused argument, but needed for different metrics within the same AnomalibMetricCollection
+            masks (torch.Tensor): Ground truth anomaly masks with non-binary values and, original height and width
+
+        Example:
+            To update the metric state for the current batch, use the ``update`` method:
+
+            >>> spro.update(preds, _, labels)
+        """
+        assert masks is not None
+        self.targets.append(masks)
+        self.preds.append(predictions)
+
+    def compute(self) -> torch.Tensor:
+        """Compute the macro average of the SPRO score across all masks in all batches.
+
+        Example:
+            To compute the metric based on the state accumulated from multiple batches, use the ``compute`` method:
+
+            >>> spro.compute()
+            tensor(0.5433)
+        """
+        targets = dim_zero_cat(self.targets)
+        preds = dim_zero_cat(self.preds)
+        return spro_score(preds, targets, threshold=self.threshold, saturation_config=self.saturation_config)
+
+
+def spro_score(
+    predictions: torch.Tensor,
+    targets: torch.Tensor,
+    threshold: float = 0.5,
+    saturation_config: dict | None = None,
+) -> torch.Tensor:
+    """Calculate the SPRO score for a batch of predictions.
+
+    Args:
+        predictions (torch.Tensor): Predicted anomaly masks
+        targets: (torch.Tensor): Ground truth anomaly masks with non-binary values and, original height and width
+        threshold (float): When predictions are passed as float, the threshold is used to binarize the predictions.
+        saturation_config (dict): Saturations configuration for each label (pixel value) as the keys.
+            Defaults: ``None`` (which the score is equivalent to PRO metric, but with the 'region' are
+            separated by mask files.
+
+    Returns:
+        torch.Tensor: Scalar value representing the average SPRO score for the input batch.
+    """
+    predictions = torch.nn.functional.interpolate(predictions.unsqueeze(1), targets.shape[1:])
+
+    # Apply threshold to binary predictions
+    if predictions.dtype == torch.float:
+        predictions = predictions > threshold
+
+    score = torch.tensor(0.0)
+    m = 0
+    # Iterate for each image in the batch
+    for i, target in enumerate(targets):
+        unique_labels = torch.unique(target)
```

Comment:
If I understand sPRO correctly, it works in the same way as PRO in a sense that each defective region counts as a separate instance. The code here assumes that each unique value is a separate instance.

So if we have a case, where let's say an image contains two regions, marked by the same value, this will not be correct, as they will be counted as one region, but should be considered as two separate instances. (I think that's why PRO implementation uses connected components to solve this issue)

Do such cases occur inside MVTec Loco? And even if they don't, what happens if datasets in future break this assumptions?

Reply:
> an image contains two regions, marked by the same value
Ah ya, good point! I missed this important point 😅. the problem occurs when we do the sum operation (merging the masks) for the case of two (or more) regions marked by the same pixel. we lose the "region" information.

Related with the overlapping issue, I think we should entirely remove the merging process, just pass the list of masks from the dataset. And then here, iterate over the list instead of over the unique labels. But this would consumes more memory i guess.

Reply:
Yes, that can happen. I think that not merging is actually a good option as you then don't have to use the connected components which is quite time-consuming. But again we'd need to verify that even a single mask doesn't have two regions with same annotation.

Reply:
i have changed the implementation, now a list of tensor is passed to the metric. seems we can resolve this conversation since the `unique_labels` approach is not used anymore.